homeassistant/components/hlk_sw16/__init__.py

```diff
@@ -2,139 +2,125 @@
-                host=host,
-                port=port,
-                disconnect_callback=disconnected,
-                reconnect_callback=reconnected,
-                loop=hass.loop,
-                timeout=CONNECTION_TIMEOUT,
-                reconnect_interval=DEFAULT_RECONNECT_INTERVAL,
-                keep_alive_interval=DEFAULT_KEEP_ALIVE_INTERVAL,
-            )
+    hass.data[DOMAIN][entry.entry_id] = {}
 
-            hass.data[DATA_DEVICE_REGISTER][device] = client
+    @callback
+    def disconnected():
+        """Schedule reconnect after connection has been lost."""
+        _LOGGER.warning("HLK-SW16 %s disconnected", address)
+        async_dispatcher_send(
+            hass, f"hlk_sw16_device_available_{entry.entry_id}", False
+        )
 
-            # Load platforms
-            hass.async_create_task(
-                async_load_platform(hass, "switch", DOMAIN, (switches, device), config)
-            )
+    @callback
+    def reconnected():
+        """Schedule reconnect after connection has been lost."""
+        _LOGGER.warning("HLK-SW16 %s connected", address)
+        async_dispatcher_send(hass, f"hlk_sw16_device_available_{entry.entry_id}", True)
+
+    async def connect():
+        """Set up connection and hook it into HA for reconnect/shutdown."""
+        _LOGGER.info("Initiating HLK-SW16 connection to %s", address)
+
+        client = await create_hlk_sw16_connection(
+            host=host,
+            port=port,
+            disconnect_callback=disconnected,
+            reconnect_callback=reconnected,
+            loop=hass.loop,
+            timeout=CONNECTION_TIMEOUT,
+            reconnect_interval=DEFAULT_RECONNECT_INTERVAL,
+            keep_alive_interval=DEFAULT_KEEP_ALIVE_INTERVAL,
+        )
 
-            # handle shutdown of HLK-SW16 asyncio transport
-            hass.bus.async_listen_once(
-                EVENT_HOMEASSISTANT_STOP, lambda x: client.stop()
-            )
+        hass.data[DOMAIN][entry.entry_id][DATA_DEVICE_REGISTER] = client
+
+        # Load entities
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, "switch")
+        )
 
-            _LOGGER.info("Connected to HLK-SW16 device: %s", device)
+        _LOGGER.info("Connected to HLK-SW16 device: %s", address)
 
-        hass.loop.create_task(connect())
+    hass.loop.create_task(connect())
 
-    for device in config[DOMAIN]:
-        add_device(device)
     return True
 
 
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    client = hass.data[DOMAIN][entry.entry_id].pop(DATA_DEVICE_REGISTER)
+    client.stop()
+    return await hass.config_entries.async_forward_entry_unload(entry, "switch")
+
+
+async def async_remove_entry(hass, entry):
+    """Stop client config entry."""
+    if hass.data[DOMAIN][entry.entry_id]:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
```

Comment:
Please move this to `async_unload_entry` as we don't want to leave this in memory when the config entry is unloaded.

You'll need to move `hass.data.setdefault(DOMAIN, {})` from `async_setup` back into `async_setup_entry` as well since deleting and readding will have called `hass.data.pop(DOMAIN)`

Reply:
moved to `async_unload_entry`